crates/sol-macro-input/src/json.rs

```diff
@@ -19,10 +19,39 @@ impl SolInput {
         let mut abi = abi.ok_or_else(|| syn::Error::new(name.span(), "ABI not found in JSON"))?;
         let sol = abi_to_sol(&name, &mut abi);
         let mut all_tokens = tokens_for_sol(&name, &sol)?.into_iter();
-        let library_tokens: TokenStream = all_tokens
+
+        let (inner_attrs, attrs) = attrs
+            .into_iter()
+            .partition::<Vec<_>, _>(|attr| matches!(attr.style, AttrStyle::Inner(_)));
+
+        let derives =
+            attrs.iter().filter(|attr| attr.path().is_ident("derive")).collect::<Vec<_>>();
+
+        let mut library_tokens_iter = all_tokens
             .by_ref()
             .take_while(|tt| !matches!(tt, TokenTree::Ident(id) if id == "interface"))
-            .collect();
+            .skip_while(|tt| matches!(tt, TokenTree::Ident(id) if id == "library"))
+            .peekable();
+
+        let library_tokens = library_tokens_iter.by_ref();
+
+        let mut libraries = Vec::new();
+
+        while library_tokens.peek().is_some() {
+            let sol_library_tokens: TokenStream = std::iter::once(TokenTree::Ident(id("library")))
+                .chain(
+                    library_tokens
+                        .take_while(|tt| !matches!(tt, TokenTree::Ident(id) if id == "library")),
```

Comment:
could this end up looping if `library_tokens.take_while` does not consume a value?

Reply:
It should just consume all left elements if closure never returns `true`

Reply:
right